---
title: "Seurat QC and Pre-Processing"
author: "Christopher O'Connor"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tictoc)
source("E:/timer_hook.R")
```

## Library

```{r message=FALSE, warning=FALSE}
# General tools
library(tidyverse)
library(dplyr)
library(patchwork) # plot tools
```

```{r library, start_timer=TRUE, name = "library_chunk", message=FALSE, warning=FALSE}
# Seurat specific tools for scRNAseq analysis
library(Seurat)
library(SeuratDisk) # save seurat objects
```

## Data Path

```{r}
# Define pathway to input file
data_path <- "C:/Users/Kenod/Pictures/count/filtered_feature_bc_matrix"
```

## Read in Data

```{r CountMatrix, start_timer=TRUE, name = "Readin_Count_Matrix_chunk"}
# Read in cellranger count aggregated matrix
count_matrix <- Read10X(data.dir = data_path)

# Remove excess variable
rm(data_path)
```

## Seurat Object

```{r SeuratObject, start_timer=TRUE, name = "Create_SeuratObj_chunk"}
# Create Seurat Object 
imm <- CreateSeuratObject(
            counts = count_matrix, 
            min.cells = 3, 
            min.genes = 200, 
            project = "imm_k23") # Define orig.ident

rm(count_matrix)
```

# 1. Quality control

## Sample Classification

```{r Sample_Classification, start_timer=TRUE, name = "Sample_Classification_chunk"}
# Sanity checked with loupe browser 
sample <- colnames(imm@assays$RNA@counts)
unique_suffix <- unique(sapply(strsplit(sample, "-"), "[", 2)) %>% print(.)

sample_suffix <- substr(sample, nchar(sample)-1, nchar(sample))
sample_suffix[sample_suffix == "-1"] <- "PI1"
sample_suffix[sample_suffix == "-2"] <- "AP1"
table(sample_suffix)
# Suffix is assigned based on run sequence provided to cellranger aggregrate
# -1 (First run) -2 (Second run)
# Therefore, it is vital to keep track of input.csv used in nextflow pipeline
# Based on this, -1 represents the Primary Pheno with Isotype treatmeent
#                -2 represents the Allograft Pheno. with anti-PD-1 treatment

imm <- AddMetaData(imm, sample_suffix, "sample")

# Remove excess values
rm(sample_suffix, unique_suffix, sample)
```

## MT genes

```{r MT, start_timer=TRUE, name = "Identify_MT_chunk"}
# Identify mito genes
mito_genes <- grep("^mt-", rownames(imm@assays$RNA@counts), value = TRUE)

# Calculate percent mito
imm$percent_mito <- PercentageFeatureSet(imm, pattern = "^mt-")

## Identify Genes per UMI (Number of distinct genes in a single cell)
# Estimate of the gene diversity/complexity of the transcriptional profile in a particular cell
imm$log10GenesPerUMI <- log10(imm$nFeature_RNA) / log10(imm$nCount_RNA) 
```

## Features

```{r}
# Feature Plot
VlnPlot(imm, features = c("nFeature_RNA", "nCount_RNA", "percent_mito"), ncol = 3)
VlnPlot(imm, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)

# QC Plot
FeatureScatter(imm, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + geom_smooth(method = 'lm')
```

```{r}
# Visualise feature-feature relationships 
FeatureScatter(imm, feature1 = "nCount_RNA", feature2 = "percent_mito")

FeatureScatter(imm, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
```

## Subset

```{r Subset, start_timer=TRUE, name = "Subset_chunk"}
# Subset with thresholds based on feature plots
imm <- subset(imm, 
               subset = nFeature_RNA > 200 & 
                        nFeature_RNA < 8000 & 
                        percent_mito < 5)
```

```{r}
# Plots after subsetting 
VlnPlot(imm, features = c("nFeature_RNA", "nCount_RNA", "percent_mito"), ncol = 3, pt.size = 0.001)
```

# 2. Pre-process Stage 1

## Normalise

```{r warning=TRUE}
# Plot before Normalisation
# Original expression distribution
raw_geneExp = as.vector(imm[['RNA']]@counts) %>% sample(10000)
raw_geneExp = raw_geneExp[raw_geneExp != 0]
hist(raw_geneExp)
```

```{r Normalisation, start_timer=TRUE, name = "Normalisation_chunk"}
# Normalise counts with log and remove unwanted cells
imm <- NormalizeData(
          object = imm, 
          normalization.method = "LogNormalize",
          scale.factor = 10000)
# NormaliseData(imm) standard line
```

```{r message=FALSE, warning=FALSE}
# After Normalisation
# New expression distribution
logNorm_geneExp = as.vector(imm[['RNA']]@data) %>% sample(10000)
logNorm_geneExp = logNorm_geneExp[logNorm_geneExp != 0]
hist(logNorm_geneExp)

rm(logNorm_geneExp, raw_geneExp)
```

## Find Variable Features

```{r Find_variable_features, start_timer=TRUE, name = "Find_variable_features_chunk", message=FALSE}
imm <- FindVariableFeatures(
          object = imm, 
          mean.function = ExpMean,
          dispersion.function = LogVMR)
```

```{r}
# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(imm), 10)
top10

# Change global options for overlapping
options(ggrepel.max.overlaps = Inf)

# FindVariableFeatures(imm, selection.method = 'vst', nfeatures = 2000)
p1 <- VariableFeaturePlot(imm)
p2 <- LabelPoints(plot = p1, points = top10, repel = TRUE)
#p1|p2
p1
p2
```

## Scale

```{r Scaling, start_timer=TRUE, name = "Scaling_chunk"}
# Scale
all.genes <- rownames(imm)
imm <- ScaleData(imm, features = all.genes)
```

```{r}
# Scale 
# imm <- ScaleData(
#          object = imm, 
#          vars.to.regress = c("nCount_RNA", "percent_mito"))

# Remove unwated sources variations through mitochondrial contamination 
# nCount_RNA the same as nUMI
```

## PCA

```{r PCA, start_timer=TRUE, name = "PCA_chunk"}
# PCA
imm <- RunPCA(
          object = imm,
          pc.genes = imm@var.genes, 
          do.print = TRUE, 
          pcs.print = 1:5, 
          genes.print = 5)
# imm <- RunPCA(imm)
```

```{r}
# visualise method 1
VizDimLoadings(imm, dims = 1:2, reduction = 'pca')
```

```{r}
# visualise method 2
DimPlot(imm, reduction = 'pca')
```

```{r}
# visualise method 3
DimHeatmap(imm, dims = 1:6, cells = 500, balanced = TRUE)
DimHeatmap(imm, dims = 6:10, cells = 500, balanced = TRUE)
```

## Determine Dimensionality

```{r}
# Method 2 instead of jackstraw
ElbowPlot(imm)
```

# 3. Pre-process Stage 2

## Neighbours

```{r Neighbours, start_timer=TRUE, name = "Neighbours_chunk"}
imm <- FindNeighbors(
          object = imm, 
          dims = 1:10)
```

## Clusters

```{r Cluster, start_timer=TRUE, name = "Clustering_chunk"}
imm <- FindClusters(
          object = imm, 
          resolution = c(0.1, 0.3, 0.5, 0.7))
```

```{r warning=FALSE}
# Loop to generate dimplots for each resolution
res_vec <- c(0.1, 0.3, 0.5, 0.7)

for(i in res_vec) {
  assign(paste0("plot_res_", i), DimPlot(imm, group.by = paste("RNA_snn_res", i, sep="."), label = T)& NoLegend())
}

# Create plot grid
plot_series_low <- plot_res_0.1 | plot_res_0.3 
plot_series_high <- plot_res_0.5 | plot_res_0.7
```

```{r}
# Print plots
plot_series_low
plot_series_high
```

```{r}
# Set cluster idents according to resolution 
Idents(imm) <- "RNA_snn_res.0.3"
```

## TSNE

```{r TSNE, start_timer=TRUE, name = "TSNE_chunk"}
# TSNE
imm <- RunTSNE(
          object = imm, 
          dims.use = 1:10, 
          do.fast = TRUE)
```

## UMAP

```{r UMAP, start_timer=TRUE, name = "UMAP_chunk"}
# UMAP
imm <- RunUMAP(
          object = imm, 
          dims = 1:10)
```

## Visualise

```{r}
DimPlot(imm, reduction = "tsne", label = T)
DimPlot(imm, reduction = "umap", label = T)
```

```{r}
DimPlot(imm, reduction = "umap", group.by = "RNA_snn_res.0.1", label = T)
DimPlot(imm, reduction = "umap", group.by = "RNA_snn_res.0.3", label = T)
```

# 4. Save Object

```{r Save_SeuratObj, start_timer=TRUE, name = "Save_SeuratObj_chunk"}
setwd("E:")
SaveH5Seurat(imm, filename = "01_k23_object.h5Seurat", overwrite = TRUE)
```

## Timer

```{r}
log.txt <- tic.log(format = TRUE)
log.txt
```
